### PR TITLE
chore: make Cast's logic reusable for other projects

### PR DESCRIPTION
**Pull request by Blizzara**
## Which issue does this PR close?

N/A

## Rationale for this change

Spark-compatible "Cast" is implemented as a physical expr in Comet, but other downstream projects might like to use it as a logical expr. To facilitate it, this makes the core logic invokable through `pub fn spark_cast`.

## What changes are included in this PR?

Makes some `Cast` struct's functions static, and moves those plus a bunch existing of static functions out of Cast impl. 
Moves ColumnarValue handling logic out of Cast::evaluate into `spark_cast`, and makes that public.
Exports `spark_cast` from spark-expr/lib

**Recommend reviewing with "Hide whitespace" ON.**

## How are these changes tested?

Existing tests

### CODE REVIEW

File: native/spark-expr/src/lib.rs [hunk 1]
```diff
@@ -24,7 +24,7 @@ mod temporal;
 pub mod timezone;
 pub mod utils;
 
-pub use cast::Cast;
+pub use cast::{Cast, spark_cast_array};
```
Comment (Blizzara) [inline native/spark-expr/src/lib.rs:27]: the export is not needed by comet, but it allows others (eg. us) to use spark_cast_array directly

File: native/spark-expr/src/cast.rs [hunk 2]
```diff
@@ -502,158 +502,163 @@ impl Cast {
...
-                // we should never reach this code because the Scala code should be checking
-                // for supported cast operations and falling back to Spark for anything that
-                // is not yet supported
-                Err(SparkError::Internal(format!(
-                    "Native cast invoked for unsupported cast from {from_type:?} to {to_type:?}"
-                )))
-            }
-        };
-        Ok(spark_cast(cast_result?, from_type, to_type))
+pub fn spark_cast_array(
```
Comment (andygrove) [inline native/spark-expr/src/cast.rs:507]: If we are making this public, we should add some documentation. I think there is a clippy lint that we should turn on to make sure all pub items have rustdocs.

Comment (Blizzara) [reply native/spark-expr/src/cast.rs]: added some in https://github.com/apache/datafusion-comet/pull/716/commits/eaf224d2ba621456942a3fd4c62c7acc2ce3b387! Though not sure what exactly would be useful here, lmk if you had anything specifc in mind :)

**Review by andygrove — APPROVED** [review]
LGTM. Thanks @Blizzara

**Pull request merged**